cassandra-launcher/src/main/scala/akka/persistence/cassandra/testkit/CassandraLauncher.scala

```diff
@@ -68,20 +70,66 @@ object CassandraLauncher {
 
   private var cassandraDaemon: Option[Closeable] = None
 
+  private val DEFAULT_HOST = "127.0.0.1"
+
   /**
    * The random free port that will be used if `port=0` is
    * specified in the `start` method.
+   *
+   * Calling `randomPort` before `start` is not recommended. It will fix the value and won't necessarily
+   * reflect the value that is effectively used by the launcher.
    */
-  lazy val randomPort: Int = freePort()
+  lazy val randomPort: Int = {
+    selectedPorts.compareAndSet((0,0), selectFreePorts(DEFAULT_HOST, 0))
+    selectedPorts.get()._1
+  }
 
+
+  @deprecated("Internal API, will be removed in future release", "0.104")
   def freePort(): Int = {
     val serverSocket = ServerSocketChannel.open().socket()
-    serverSocket.bind(new InetSocketAddress("127.0.0.1", 0))
+    serverSocket.bind(new InetSocketAddress(DEFAULT_HOST, 0))
     val port = serverSocket.getLocalPort
     serverSocket.close()
     port
   }
 
+
+  private val selectedPorts: AtomicReference[(Int, Int)] = new AtomicReference((0,0))
+
+  /**
+   * Select two free ports.
+   * Note that requestPort is always used even if user requested a fixed port. We want to make sure the port it no in use
+   * and won't conflict with the storagePort
+   */
+  private def selectFreePorts(host: String, requestedPort: Int): (Int, Int) = {
+
+    val clientSocket = ServerSocketChannel.open().socket()
+    val storageSocket = ServerSocketChannel.open().socket()
+
+    try {
+      clientSocket.bind(new InetSocketAddress(host, requestedPort))
+      storageSocket.bind(new InetSocketAddress(host, 0))
+
+      // return both ports
+      (clientSocket.getLocalPort, storageSocket.getLocalPort)
+
+    } finally  {
+      // close independently
+      val t1 = Try(clientSocket.close())
+      val t2 = Try(storageSocket.close())
+
+      // if one the two failed, we should throw the exception
+      (t1, t2) match {
+        case (Failure(ex1), Failure(ex2)) => throw new RuntimeException(s"Failed to close sockets: client '${ex1.getMessage}', storage '${ex2.getMessage}'" )
+        case (Failure(ex1), _) => throw new RuntimeException(s"Failed to close client-port socket: '${ex1.getMessage}'" )
+        case (_, Failure(ex2)) => throw new RuntimeException(s"Failed to close storage-port socket: '${ex2.getMessage}'" )
+        case (_, _) => // we are fine, all closed
```

Comment:
Not so sure if really important to do this. 

Maybe 

```scala
t1.flatMap(_ => t2).get 
```

would be enough (?).